File: go/internal/movegen/queen.go
package movegen

import (
	"github.com/notmalte/psce/internal/bitboard"
	"github.com/notmalte/psce/internal/constants"
	"github.com/notmalte/psce/internal/move"
	"github.com/notmalte/psce/internal/position"
	"math/bits"
)

type QueenMoveGen struct {
	rookMoveGen   *RookMoveGen
	bishopMoveGen *BishopMoveGen
}

func (qmg *QueenMoveGen) GetAttacks(square uint8, occupancy uint64) uint64 {
	return qmg.rookMoveGen.GetAttacks(square, occupancy) | qmg.bishopMoveGen.GetAttacks(square, occupancy)
}

func NewQueenMoveGen(rmg *RookMoveGen, bmg *BishopMoveGen) *QueenMoveGen {
	return &QueenMoveGen{
		rookMoveGen:   rmg,
		bishopMoveGen: bmg,
	}
}

func (qmg *QueenMoveGen) GeneratePseudoLegalMoves(pos *position.Position) []move.Move {
	isWhite := pos.ColorToMove == constants.ColorWhite

	var otherColor uint8
	var piece uint8
	if isWhite {
		otherColor = constants.ColorBlack
		piece = constants.WhiteQueen
	} else {
		otherColor = constants.ColorWhite
		piece = constants.BlackQueen
	}

	moves := []move.Move{}
	bb := pos.PieceBitboards[piece]

	for bb != 0 {
		fromSquare := uint8(bits.TrailingZeros64(bb))

		attacks := qmg.GetAttacks(fromSquare, pos.ColorBitboards[constants.ColorBoth]) & ^pos.ColorBitboards[pos.ColorToMove]

		for attacks != 0 {
			toSquare := uint8(bits.TrailingZeros64(attacks))

			flags := constants.MoveFlagNone
			if bitboard.GetBit(pos.ColorBitboards[otherColor], toSquare) {
				flags = constants.MoveFlagCapture
			}

			moves = append(moves, move.Move{
				FromSquare: fromSquare,
				ToSquare:   toSquare,
				Piece:      piece,
				Flags:      flags,
			})

			bitboard.ClearBit(&attacks, toSquare)
		}

		bitboard.ClearBit(&bb, fromSquare)
	}

	return moves
}

File: src/engine/movegen/queen.rs
use crate::engine::{
    bitboard::{Bitboard, Square},
    color::Color,
    movegen::{bishop::BishopMoveGen, rook::RookMoveGen},
    moves::{Move, MoveFlags},
    piece::Piece,
    position::Position,
};

pub struct QueenMoveGen {
    bishop: BishopMoveGen,
    rook: RookMoveGen,
}

impl QueenMoveGen {
    pub fn new(bishop: BishopMoveGen, rook: RookMoveGen) -> Self {
        Self { bishop, rook }
    }

    fn get_attacks(&self, square: Square, occupancy: Bitboard) -> Bitboard {
        self.bishop.get_attacks(square, occupancy) | self.rook.get_attacks(square, occupancy)
    }

    pub fn generate_moves(&self, position: &Position) -> Vec<Move> {
        let color = position.color_to_move();

        let piece = match color {
            Color::White => Piece::WhiteQueen,
            Color::Black => Piece::BlackQueen,
        };

        let mut moves = vec![];

        let queens = position.bitboards().piece(piece);

        for from_square in queens.squares() {
            let attacks = self.get_attacks(from_square, position.bitboards().all())
                & !position.bitboards().color(color);

            for to_square in attacks.squares() {
                let capture = position.bitboards().all().get(to_square);

                let flags = if capture {
                    MoveFlags::CAPTURE
                } else {
                    MoveFlags::NONE
                };

                moves.push(Move::new(from_square, to_square, piece, None, flags));
            }
        }

        moves
    }
}
